## Branch and Bound Example

Consider the following **integer programming problem**:

\begin{aligned}
\max \quad & z = 18x_e + 6x_f \\
\text{s.t.} \quad 
& x_e - 3x_f \leq 0 \\
& 42.8x_e + 100x_f \leq 800 \\
& 20x_e + 6x_f \leq 142 \\
& 30x_e + 10x_f \geq 135 \\
& x_e, x_f \geq 0 \quad \text{and integer}
\end{aligned}


### Exercise

1. Solve the **LP relaxation** of the problem (ignore the integer requirement).
2. Examine whether the solution is **integer feasible**.
3. If not, proceed with the **Branch and Bound** algorithm:
   - Which variable would you branch on?
   - What subproblems would be created?

In [1]:
import gurobipy as gp
from gurobipy import GRB

### Step 1: Solve the LP relaxation

In [2]:
# Create model
model = gp.Model("Branch_and_Bound_Example")

# Decision variables
x_e = model.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = model.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)

# Objective function
model.setObjective(18 * x_e + 6 * x_f, GRB.MAXIMIZE)

# Constraints
model.addConstr(x_e - 3 * x_f <= 0, name="c1")
model.addConstr(42.8 * x_e + 100 * x_f <= 800, name="c2")
model.addConstr(20 * x_e + 6 * x_f <= 142, name="c3")
model.addConstr(30 * x_e + 10 * x_f >= 135, name="c4")

# Optimize
model.optimize()

# Print solution
if model.status == GRB.OPTIMAL:
    print("\nOptimal Solution:")
    print(f"x_e = {x_e.X}")
    print(f"x_f = {x_f.X}")
    print(f"Optimal objective value = {model.objVal}")

Restricted license - for non-production use only - expires 2027-11-29
Gurobi Optimizer version 13.0.1 build v13.0.1rc0 (mac64[arm] - Darwin 24.6.0 24G517)

CPU model: Apple M4 Pro
Thread count: 14 physical cores, 14 logical processors, using up to 14 threads

Optimize a model with 4 rows, 2 columns and 8 nonzeros (Max)
Model fingerprint: 0x4f2323b9
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [1e+00, 1e+02]
  Objective range  [6e+00, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+02, 8e+02]

Presolve removed 4 rows and 2 columns
Presolve time: 0.01s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.3121524e+02   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.01 seconds (0.00 work units)
Optimal objective  1.312152363e+02

Optimal Solution:
x_e = 5.39238182652593
x_f = 5.6920605782469025
Optimal objective value = 131.21523634694813


Examine whether the solution is integer feasible. If not, use the branch and bound algorithm!

### Step 2: Branch on the decision variable with the largest decimal

Solve Sub Problem 1:

\begin{aligned}
\max \quad & z = 18x_e + 6x_f \\
\text{s.t.} \quad 
& x_e - 3x_f \leq 0 \\
& 42.8x_e + 100x_f \leq 800 \\
& 20x_e + 6x_f \leq 142 \\
& 30x_e + 10x_f \geq 135 \\
& x_f \geq 6 \\
& x_e, x_f \geq 0 
\end{aligned}

In [3]:
# Create the sub problem 1
sub_m1 = gp.Model("Branch_and_Bound_Example")

# Decision variables
x_e = sub_m1.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = sub_m1.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)

# Objective function
sub_m1.setObjective(18 * x_e + 6 * x_f, GRB.MAXIMIZE)

# Constraints
sub_m1.addConstr(x_e - 3 * x_f <= 0, name="c1")
sub_m1.addConstr(42.8 * x_e + 100 * x_f <= 800, name="c2")
sub_m1.addConstr(20 * x_e + 6 * x_f <= 142, name="c3")
sub_m1.addConstr(30 * x_e + 10 * x_f >= 135, name="c4")
sub_m1.addConstr(x_f >= 6, name="c5")

# Optimize
sub_m1.optimize()

# Print solution
if sub_m1.status == GRB.OPTIMAL:
    print("\nOptimal Solution:")
    print(f"x_e = {x_e.X}")
    print(f"x_f = {x_f.X}")
    print(f"Optimal objective value = {sub_m1.objVal}")

Gurobi Optimizer version 13.0.1 build v13.0.1rc0 (mac64[arm] - Darwin 24.6.0 24G517)

CPU model: Apple M4 Pro
Thread count: 14 physical cores, 14 logical processors, using up to 14 threads

Optimize a model with 5 rows, 2 columns and 9 nonzeros (Max)
Model fingerprint: 0xc6ab98e7
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [1e+00, 1e+02]
  Objective range  [6e+00, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [6e+00, 8e+02]

Presolve removed 5 rows and 2 columns
Presolve time: 0.00s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.2011215e+02   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  1.201121495e+02

Optimal Solution:
x_e = 4.672897196261682
x_f = 6.0
Optimal objective value = 120.11214953271028


Examine whether the solution is integer feasible. If not, continue using the branch and bound algorithm!

Solve Sub Problem 2:

\begin{aligned}
\max \quad & z = 18x_e + 6x_f \\
\text{s.t.} \quad 
& x_e - 3x_f \leq 0 \\
& 42.8x_e + 100x_f \leq 800 \\
& 20x_e + 6x_f \leq 142 \\
& 30x_e + 10x_f \geq 135 \\
& x_f \leq 5 \\
& x_e, x_f \geq 0 
\end{aligned}

In [4]:
# Create the sub problem 2
sub_m2 = gp.Model("Branch_and_Bound_Example")

# Decision variables
x_e = sub_m2.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = sub_m2.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)

# Objective function
sub_m2.setObjective(18 * x_e + 6 * x_f, GRB.MAXIMIZE)

# Constraints
sub_m2.addConstr(x_e - 3 * x_f <= 0, name="c1")
sub_m2.addConstr(42.8 * x_e + 100 * x_f <= 800, name="c2")
sub_m2.addConstr(20 * x_e + 6 * x_f <= 142, name="c3")
sub_m2.addConstr(30 * x_e + 10 * x_f >= 135, name="c4")
sub_m2.addConstr(x_f <= 5, name="c6")

# Optimize
sub_m2.optimize()

# Print solution
if sub_m2.status == GRB.OPTIMAL:
    print("\nOptimal Solution:")
    print(f"x_e = {x_e.X}")
    print(f"x_f = {x_f.X}")
    print(f"Optimal objective value = {sub_m2.objVal}")

Gurobi Optimizer version 13.0.1 build v13.0.1rc0 (mac64[arm] - Darwin 24.6.0 24G517)

CPU model: Apple M4 Pro
Thread count: 14 physical cores, 14 logical processors, using up to 14 threads

Optimize a model with 5 rows, 2 columns and 9 nonzeros (Max)
Model fingerprint: 0xc5748809
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [1e+00, 1e+02]
  Objective range  [6e+00, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+00, 8e+02]

Presolve removed 5 rows and 2 columns
Presolve time: 0.00s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.3080000e+02   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  1.308000000e+02

Optimal Solution:
x_e = 5.6
x_f = 5.0
Optimal objective value = 130.8


Sub Problem 2 returned a better optimal solution than Sub Problem 1; therefore, we will continue branching on Sub Problem 2 first!

Solve Sub Problem 3:

\begin{aligned}
\max \quad & z = 18x_e + 6x_f \\
\text{s.t.} \quad 
& x_e - 3x_f \leq 0 \\
& 42.8x_e + 100x_f \leq 800 \\
& 20x_e + 6x_f \leq 142 \\
& 30x_e + 10x_f \geq 135 \\
& x_f \leq 5 \\
& x_e \geq 6 \\
& x_e, x_f \geq 0 
\end{aligned}

In [5]:
# Create the sub problem 3
sub_m3 = gp.Model("Branch_and_Bound_Example")

# Decision variables
x_e = sub_m3.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = sub_m3.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)

# Objective function
sub_m3.setObjective(18 * x_e + 6 * x_f, GRB.MAXIMIZE)

# Constraints
sub_m3.addConstr(x_e - 3 * x_f <= 0, name="c1")
sub_m3.addConstr(42.8 * x_e + 100 * x_f <= 800, name="c2")
sub_m3.addConstr(20 * x_e + 6 * x_f <= 142, name="c3")
sub_m3.addConstr(30 * x_e + 10 * x_f >= 135, name="c4")
sub_m3.addConstr(x_f <= 5, name="c6")
sub_m3.addConstr(x_e >= 6, name="c7")

# Optimize
sub_m3.optimize()

# Print solution
if sub_m3.status == GRB.OPTIMAL:
    print("\nOptimal Solution:")
    print(f"x_e = {x_e.X}")
    print(f"x_f = {x_f.X}")
    print(f"Optimal objective value = {sub_m3.objVal}")

Gurobi Optimizer version 13.0.1 build v13.0.1rc0 (mac64[arm] - Darwin 24.6.0 24G517)

CPU model: Apple M4 Pro
Thread count: 14 physical cores, 14 logical processors, using up to 14 threads

Optimize a model with 6 rows, 2 columns and 10 nonzeros (Max)
Model fingerprint: 0x1988c151
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [1e+00, 1e+02]
  Objective range  [6e+00, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+00, 8e+02]

Presolve removed 6 rows and 2 columns
Presolve time: 0.00s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.3000000e+02   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  1.300000000e+02

Optimal Solution:
x_e = 6.0
x_f = 3.6666666666666665
Optimal objective value = 130.0


Solve Sub Problem 4:

\begin{aligned}
\max \quad & z = 18x_e + 6x_f \\
\text{s.t.} \quad 
& x_e - 3x_f \leq 0 \\
& 42.8x_e + 100x_f \leq 800 \\
& 20x_e + 6x_f \leq 142 \\
& 30x_e + 10x_f \geq 135 \\
& x_f \leq 5 \\
& x_e \leq 5 \\
& x_e, x_f \geq 0 
\end{aligned}

In [6]:
# Create the sub problem 4
sub_m4 = gp.Model("Branch_and_Bound_Example")

# Decision variables
x_e = sub_m4.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = sub_m4.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)

# Objective function
sub_m4.setObjective(18 * x_e + 6 * x_f, GRB.MAXIMIZE)

# Constraints
sub_m4.addConstr(x_e - 3 * x_f <= 0, name="c1")
sub_m4.addConstr(42.8 * x_e + 100 * x_f <= 800, name="c2")
sub_m4.addConstr(20 * x_e + 6 * x_f <= 142, name="c3")
sub_m4.addConstr(30 * x_e + 10 * x_f >= 135, name="c4")
sub_m4.addConstr(x_f <= 5, name="c6")
sub_m4.addConstr(x_e <= 5, name="c8")

# Optimize
sub_m4.optimize()

# Print solution
if sub_m4.status == GRB.OPTIMAL:
    print("\nOptimal Solution:")
    print(f"x_e = {x_e.X}")
    print(f"x_f = {x_f.X}")
    print(f"Optimal objective value = {sub_m4.objVal}")

Gurobi Optimizer version 13.0.1 build v13.0.1rc0 (mac64[arm] - Darwin 24.6.0 24G517)

CPU model: Apple M4 Pro
Thread count: 14 physical cores, 14 logical processors, using up to 14 threads

Optimize a model with 6 rows, 2 columns and 10 nonzeros (Max)
Model fingerprint: 0x168cd073
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [1e+00, 1e+02]
  Objective range  [6e+00, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+00, 8e+02]

Presolve removed 6 rows and 2 columns
Presolve time: 0.00s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.2000000e+02   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  1.200000000e+02

Optimal Solution:
x_e = 5.0
x_f = 5.0
Optimal objective value = 120.0


---
Sub Problem 4 has given us an integer solution of 120; we can stop here for this branch! However, we still need to continue branching on Sub Problem 3, which has given us: 

<br> Optimal Solution:
<br> $x_e$ = 6.0
<br> $x_f$ = 3.666
<br> Optimal objective value = 130.0

---

Solve Sub Problem 5:

\begin{aligned}
\max \quad & z = 18x_e + 6x_f \\
\text{s.t.} \quad 
& x_e - 3x_f \leq 0 \\
& 42.8x_e + 100x_f \leq 800 \\
& 20x_e + 6x_f \leq 142 \\
& 30x_e + 10x_f \geq 135 \\
& x_f \leq 5 \\
& x_e \geq 6 \\
& x_f \leq 3 \\
& x_e, x_f \geq 0 
\end{aligned}

In [7]:
# Create the sub problem 5
sub_m5 = gp.Model("Branch_and_Bound_Example")

# Decision variables
x_e = sub_m5.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = sub_m5.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)

# Objective function
sub_m5.setObjective(18 * x_e + 6 * x_f, GRB.MAXIMIZE)

# Constraints
sub_m5.addConstr(x_e - 3 * x_f <= 0, name="c1")
sub_m5.addConstr(42.8 * x_e + 100 * x_f <= 800, name="c2")
sub_m5.addConstr(20 * x_e + 6 * x_f <= 142, name="c3")
sub_m5.addConstr(30 * x_e + 10 * x_f >= 135, name="c4")
sub_m5.addConstr(x_f <= 5, name="c6")
sub_m5.addConstr(x_e >= 6, name="c7")
sub_m5.addConstr(x_f <= 3, name="c9")4

# Optimize
sub_m5.optimize()

# Print solution
if sub_m5.status == GRB.OPTIMAL:
    print("\nOptimal Solution:")
    print(f"x_e = {x_e.X}")
    print(f"x_f = {x_f.X}")
    print(f"Optimal objective value = {sub_m5.objVal}")

SyntaxError: invalid syntax (918488548.py, line 18)

Solve Sub Problem 6:

\begin{aligned}
\max \quad & z = 18x_e + 6x_f \\
\text{s.t.} \quad 
& x_e - 3x_f \leq 0 \\
& 42.8x_e + 100x_f \leq 800 \\
& 20x_e + 6x_f \leq 142 \\
& 30x_e + 10x_f \geq 135 \\
& x_f \leq 5 \\
& x_e \geq 6 \\
& x_f \geq 4 \\
& x_e, x_f \geq 0 
\end{aligned}

In [ ]:
# Create the sub problem 6
sub_m6 = gp.Model("Branch_and_Bound_Example")

# Decision variables
x_e = sub_m6.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = sub_m6.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)

# Objective function
sub_m6.setObjective(18 * x_e + 6 * x_f, GRB.MAXIMIZE)

# Constraints
sub_m6.addConstr(x_e - 3 * x_f <= 0, name="c1")
sub_m6.addConstr(42.8 * x_e + 100 * x_f <= 800, name="c2")
sub_m6.addConstr(20 * x_e + 6 * x_f <= 142, name="c3")
sub_m6.addConstr(30 * x_e + 10 * x_f >= 135, name="c4")
sub_m6.addConstr(x_f <= 5, name="c6")
sub_m6.addConstr(x_e >= 6, name="c7")
sub_m6.addConstr(x_f >= 4, name="c10")

# Optimize
sub_m6.optimize()

# Print solution
if sub_m6.status == GRB.OPTIMAL:
    print("\nOptimal Solution:")
    print(f"x_e = {x_e.X}")
    print(f"x_f = {x_f.X}")
    print(f"Optimal objective value = {sub_m6.objVal}")

Gurobi Optimizer version 10.0.3 build v10.0.3rc0 (win64)

CPU model: 13th Gen Intel(R) Core(TM) i9-13900H, instruction set [SSE2|AVX|AVX2]
Thread count: 14 physical cores, 20 logical processors, using up to 20 threads

Optimize a model with 7 rows, 2 columns and 11 nonzeros
Model fingerprint: 0x6990990f
Coefficient statistics:
  Matrix range     [1e+00, 1e+02]
  Objective range  [6e+00, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [4e+00, 8e+02]
Presolve removed 4 rows and 0 columns
Presolve time: 0.01s

Solved in 0 iterations and 0.01 seconds (0.00 work units)
Infeasible or unbounded model


---
Sub Problem 6 gave us an infeasible solution! We can prune this node as well! But we need to continue on Sub-Problem 5 which gave us the solution:
    
<br> Optimal Solution:
<br> $x_e$ = 6.2
<br> $x_f$ = 3.0
<br> Optimal objective value = 129.60   

---

Solve Sub Problem 7:

\begin{aligned}
\max \quad & z = 18x_e + 6x_f \\
\text{s.t.} \quad 
& x_e - 3x_f \leq 0 \\
& 42.8x_e + 100x_f \leq 800 \\
& 20x_e + 6x_f \leq 142 \\
& 30x_e + 10x_f \geq 135 \\
& x_f \leq 5 \\
& x_e \geq 6 \\
& x_f \leq 3 \\
& x_e \leq 6 \\
& x_e, x_f \geq 0 
\end{aligned}

In [ ]:
# Create the sub problem 7
sub_m7 = gp.Model("Branch_and_Bound_Example")

# Decision variables
x_e = sub_m7.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = sub_m7.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)

# Objective function
sub_m7.setObjective(18 * x_e + 6 * x_f, GRB.MAXIMIZE)

# Constraints
sub_m7.addConstr(x_e - 3 * x_f <= 0, name="c1")
sub_m7.addConstr(42.8 * x_e + 100 * x_f <= 800, name="c2")
sub_m7.addConstr(20 * x_e + 6 * x_f <= 142, name="c3")
sub_m7.addConstr(30 * x_e + 10 * x_f >= 135, name="c4")
sub_m7.addConstr(x_f <= 5, name="c6")
sub_m7.addConstr(x_e >= 6, name="c7")
sub_m7.addConstr(x_f <= 3, name="c9")
sub_m7.addConstr(x_e <= 6, name="c11")

# Optimize
sub_m7.optimize()

# Print solution
if sub_m7.status == GRB.OPTIMAL:
    print("\nOptimal Solution:")
    print(f"x_e = {x_e.X}")
    print(f"x_f = {x_f.X}")
    print(f"Optimal objective value = {sub_m7.objVal}")

Gurobi Optimizer version 10.0.3 build v10.0.3rc0 (win64)

CPU model: 13th Gen Intel(R) Core(TM) i9-13900H, instruction set [SSE2|AVX|AVX2]
Thread count: 14 physical cores, 20 logical processors, using up to 20 threads

Optimize a model with 8 rows, 2 columns and 12 nonzeros
Model fingerprint: 0xc58d5404
Coefficient statistics:
  Matrix range     [1e+00, 1e+02]
  Objective range  [6e+00, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+00, 8e+02]
Presolve removed 8 rows and 2 columns
Presolve time: 0.01s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.2600000e+02   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.02 seconds (0.00 work units)
Optimal objective  1.260000000e+02

Optimal Solution:
x_e = 6.0
x_f = 3.0
Optimal objective value = 126.0


Sub-Problem 7 gave us an integer optimal solution, we can stop branching from here as well! Let's check the other side of the branch to see whether it will result in a better optimal solution or not. 

---

Solve Sub Problem 8:

\begin{aligned}
\max \quad & z = 18x_e + 6x_f \\
\text{s.t.} \quad 
& x_e - 3x_f \leq 0 \\
& 42.8x_e + 100x_f \leq 800 \\
& 20x_e + 6x_f \leq 142 \\
& 30x_e + 10x_f \geq 135 \\
& x_f \leq 5 \\
& x_e \geq 6 \\
& x_f \leq 3 \\
& x_e \geq 7 \\
& x_e, x_f \geq 0 
\end{aligned}

In [ ]:
# Create the sub problem 8
sub_m8 = gp.Model("Branch_and_Bound_Example")

# Decision variables
x_e = sub_m8.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = sub_m8.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)

# Objective function
sub_m8.setObjective(18 * x_e + 6 * x_f, GRB.MAXIMIZE)

# Constraints
sub_m8.addConstr(x_e - 3 * x_f <= 0, name="c1")
sub_m8.addConstr(42.8 * x_e + 100 * x_f <= 800, name="c2")
sub_m8.addConstr(20 * x_e + 6 * x_f <= 142, name="c3")
sub_m8.addConstr(30 * x_e + 10 * x_f >= 135, name="c4")
sub_m8.addConstr(x_f <= 5, name="c6")
sub_m8.addConstr(x_e >= 6, name="c7")
sub_m8.addConstr(x_f <= 3, name="c9")
sub_m8.addConstr(x_e >= 7, name="c12")

# Optimize
sub_m8.optimize()

# Print solution
if sub_m8.status == GRB.OPTIMAL:
    print("\nOptimal Solution:")
    print(f"x_e = {x_e.X}")
    print(f"x_f = {x_f.X}")
    print(f"Optimal objective value = {sub_m8.objVal}")

Gurobi Optimizer version 10.0.3 build v10.0.3rc0 (win64)

CPU model: 13th Gen Intel(R) Core(TM) i9-13900H, instruction set [SSE2|AVX|AVX2]
Thread count: 14 physical cores, 20 logical processors, using up to 20 threads

Optimize a model with 8 rows, 2 columns and 12 nonzeros
Model fingerprint: 0xfbbaa786
Coefficient statistics:
  Matrix range     [1e+00, 1e+02]
  Objective range  [6e+00, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+00, 8e+02]
Presolve removed 5 rows and 0 columns
Presolve time: 0.01s

Solved in 0 iterations and 0.01 seconds (0.00 work units)
Infeasible or unbounded model


We still did not branch out on Sub-Problem 1; however, there's no need as the best solution it can return will not be higher than the current integer optimal solution of 126!

## 常见疑惑：为什么LP最优值可以作为分支剪枝的依据？

### 您的疑问
为什么当右分支的LP最优值（比如4.7）小于左分支的LP最优值（比如5.6）时，我们就能直接剪掉右分支？毕竟左分支的LP最优本身不是整数，能保证左分支的整数最优一定比右分支好吗？

### 关键概念澄清：**上界性质**

这里的核心在于理解：**LP最优值是该分支中任何整数可行解的上界**

让我用一个具体例子来说明：

**假设原问题是最大化问题：**
```
max  Z = 3x_e + 2x_f
s.t. x_e + x_f ≤ 10
     x_e, x_f ≥ 0, 整数
```

**原始LP松弛（无分支）：**
- LP最优：(10, 0) 或 (0, 10) 等，Z = 30（但这里假设为16）
- LP最优解：x_e = 4.7, x_f = 5.3，Z = 5.6

**分支1：x_f ≤ 5**
- 添加约束：x_f ≤ 5
- 新的LP最优：x_e = 4.7, x_f = 5，Z = 18.1
- **关键：这个分支中，任何整数解都不能超过18.1**

**分支2：x_f ≥ 6**
- 添加约束：x_f ≥ 6
- 新的LP最优：x_e = 4, x_f = 6，Z = 16.0
- **关键：这个分支中，任何整数解都不能超过16.0**

**当前找到的最佳整数解：Z = 17.0**

现在比较：
- 分支1的上界 = 18.1 > 17.0 → **需要继续探索**（可能找到更优解）
- 分支2的上界 = 16.0 < 17.0 → **可以剪掉**（不可能找到比17.0更优的解）

### 为什么这个逻辑是对的？

**数学证明：**

设某分支的LP最优值为 $Z_{LP}$，该分支的约束集合为 $S$。

对于任意整数可行解 $(x_e^*, x_f^*) \in S$，必然满足：
- $(x_e^*, x_f^*)$ 是整数
- $(x_e^*, x_f^*)$ 满足该分支的所有约束

因为LP是原IP去掉整数约束，所以：
$$Z(x_e^*, x_f^*) \leq Z_{LP}$$

这是因为 LP 可行域包含了 IP 的所有可行解，所以 LP 最优值 ≥ 该约束下任何整数解的目标值。

### 您例子的分析

您说"左边(5,5) optimal 比右边的取整数值更优"，这里有个关键点：

如果 (5,5) 是左分支（x_f ≤ 5）的可行解，那么它一定被左分支的LP最优值 bound 住：
$$Z(5,5) \leq Z_{LP}^{左} = 5.6$$

所以：
- 如果左分支LP最优值 = 5.6
- 左分支的任何整数解（包括(5,5)）的目标值 ≤ 5.6
- 这意味着 Z(5,5) ≤ 5.6

如果您看到 Z(5,5) 的实际值是 4.8，那确实 < 5.6，符合这个上界。

### 总结

B&B不是比较LP最优值本身的好坏，而是比较**上界的紧度**：
- LP最优值 = 该分支所有可行解（包括非整数和整数）能达到的最大值的上界
- 如果该上界 < 当前找到的最佳整数解，就不可能在这个分支找到更优解
- 这样就能安全地剪掉该分支


In [8]:
import numpy as np
import pandas as pd
from gurobipy import *

# 一个简单的例子来演示LP上界如何工作
print("=" * 80)
print("演示：为什么LP最优值是分支剪枝的有效依据")
print("=" * 80)

# 原始问题
print("\n【原始IP问题】")
print("max Z = 3*x_e + 2*x_f")
print("s.t. x_e + 2*x_f ≤ 12")
print("     x_e, x_f ≥ 0, 整数")

# 原始LP松弛
m0 = Model("Original LP")
x_e = m0.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f = m0.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)
m0.setObjective(3*x_e + 2*x_f, GRB.MAXIMIZE)
m0.addConstr(x_e + 2*x_f <= 12)
m0.optimize()

print(f"\n【原始LP松弛解】")
print(f"  x_e = {x_e.X:.2f}, x_f = {x_f.X:.2f}")
print(f"  Z_LP = {m0.ObjVal:.2f}")
print(f"  → 这是一个上界：任何整数解都不能超过 {m0.ObjVal:.2f}")

# 分支1：x_f ≤ 5
print(f"\n【分支1：x_f ≤ 5】")
m1 = Model("Branch 1: x_f <= 5")
x_e1 = m1.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f1 = m1.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0, ub=5)
m1.setObjective(3*x_e1 + 2*x_f1, GRB.MAXIMIZE)
m1.addConstr(x_e1 + 2*x_f1 <= 12)
m1.optimize()

print(f"  x_e = {x_e1.X:.2f}, x_f = {x_f1.X:.2f}")
print(f"  Z_LP(分支1) = {m1.ObjVal:.2f}")
print(f"  → 这个分支的上界是 {m1.ObjVal:.2f}")

# 分支2：x_f ≥ 6
print(f"\n【分支2：x_f ≥ 6】")
m2 = Model("Branch 2: x_f >= 6")
x_e2 = m2.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f2 = m2.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=6)
m2.setObjective(3*x_e2 + 2*x_f2, GRB.MAXIMIZE)
m2.addConstr(x_e2 + 2*x_f2 <= 12)
m2.optimize()

print(f"  x_e = {x_e2.X:.2f}, x_f = {x_f2.X:.2f}")
print(f"  Z_LP(分支2) = {m2.ObjVal:.2f}")
print(f"  → 这个分支的上界是 {m2.ObjVal:.2f}")

# 现在找整数解
print(f"\n【寻找整数解】")
m_int = Model("Original IP")
x_e_int = m_int.addVar(vtype=GRB.INTEGER, name="x_e", lb=0)
x_f_int = m_int.addVar(vtype=GRB.INTEGER, name="x_f", lb=0)
m_int.setObjective(3*x_e_int + 2*x_f_int, GRB.MAXIMIZE)
m_int.addConstr(x_e_int + 2*x_f_int <= 12)
m_int.optimize()

print(f"  整数最优解：x_e = {x_e_int.X:.0f}, x_f = {x_f_int.X:.0f}")
print(f"  整数最优值：Z* = {m_int.ObjVal:.2f}")

# 比较上界与整数解
print(f"\n【剪枝决策】")
print(f"  分支1上界：{m1.ObjVal:.2f}")
print(f"  分支2上界：{m2.ObjVal:.2f}")
print(f"  当前最优整数解：Z* = {m_int.ObjVal:.2f}")
print(f"\n  分支1: {m1.ObjVal:.2f} {'> Z*，需要继续探索' if m1.ObjVal > m_int.ObjVal else '≤ Z*，可以剪掉'}")
print(f"  分支2: {m2.ObjVal:.2f} {'> Z*，需要继续探索' if m2.ObjVal > m_int.ObjVal else '≤ Z*，可以剪掉'}")

# 验证：列出分支1中的所有整数解
print(f"\n【验证：分支1中所有可行的整数解】")
solutions = []
for x_e_val in range(13):
    for x_f_val in range(6):  # x_f ≤ 5
        if x_e_val + 2*x_f_val <= 12:
            obj_val = 3*x_e_val + 2*x_f_val
            solutions.append({
                'x_e': x_e_val,
                'x_f': x_f_val,
                'Z': obj_val,
                '满足上界': obj_val <= m1.ObjVal
            })

df = pd.DataFrame(solutions).sort_values('Z', ascending=False)
print(f"  分支1中Z最大的整数解：{df.iloc[0]}")
print(f"  它的目标值 {df.iloc[0]['Z']} ≤ 分支1的LP上界 {m1.ObjVal:.2f} ✓")
print(f"\n  分支1中所有整数解的Z都满足 ≤ {m1.ObjVal:.2f}")
print(f"  → 这验证了LP上界的有效性！")


演示：为什么LP最优值是分支剪枝的有效依据

【原始IP问题】
max Z = 3*x_e + 2*x_f
s.t. x_e + 2*x_f ≤ 12
     x_e, x_f ≥ 0, 整数
Gurobi Optimizer version 13.0.1 build v13.0.1rc0 (mac64[arm] - Darwin 24.6.0 24G517)

CPU model: Apple M4 Pro
Thread count: 14 physical cores, 14 logical processors, using up to 14 threads

Optimize a model with 1 rows, 2 columns and 2 nonzeros (Max)
Model fingerprint: 0xe4d2dbf0
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [1e+00, 2e+00]
  Objective range  [2e+00, 3e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+01, 1e+01]

Presolve removed 1 rows and 2 columns
Presolve time: 0.00s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    3.6000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  3.600000000e+01

【原始LP松弛解】
  x_e = 12.00, x_f = 0.00
  Z_LP = 36.00
  → 这是一个上界：任何整数解都不能超过 36.00

【分支1：x_f ≤ 5】
Gur

### 深入理解：为什么(5,5)必然被左分支LP上界约束

您提到的例子："左边(5,5)的最优值可能小于右边某个整数解"。让我用数学来说明这为什么不会导致问题：

**给定：**
- 左分支约束：x_f ≤ 5
- 左分支LP最优值：Z_LP^左 = 5.6，最优点：(4.7, 5.3)
- 候选整数点：(5, 5)

**关键数学性质：**

如果 (5, 5) 在左分支的可行域内（即满足 x_f ≤ 5 和其他约束），那么由于 LP 的可行域包含了所有整数点的凸包：

$$Z(5,5) \leq Z_{LP}^{左} = 5.6$$

这里用到的是一个基本的线性规划性质：**任何可行点的目标函数值都不会超过 LP 最优值**。

**为什么？**因为 LP 的最优值是在所有可行点（包括整数点及其凸包）中的最大值。

### 一个可能的困惑点

您可能在想："但 objective function 不是单调的啊，凸包内可能有更优的非整数点，为什么整数点一定被 LP 最优值限制？"

答案是：**这不是关于单调性或凸性的问题，而是关于线性规划的基本性质**。

- 即使目标函数不是凸的（但它本身就是线性的，所以自动是凸的）
- 即使可行域的形状很复杂
- 对于任何线性目标函数，该目标函数值在该约束下的最大值就是 LP 最优值

**简单证明：**
设整数可行解为 $(x^*)$，它满足约束 $S$（包括 x_f ≤ 5）。
对于线性目标函数 $c^T x$：

$$c^T x^* \leq \max_{x \in S} c^T x = Z_{LP}^{左}$$

这是定义，不需要任何额外的假设！

### 您的例子为什么不会出现问题

假设您的例子中：
- 右分支（x_f ≥ 6）的LP最优值 = 4.7
- 左分支（x_f ≤ 5）的LP最优值 = 5.6
- 左分支的整数最优解(5,5)的目标值 = 4.8
- 右分支的整数最优解(4,6)的目标值 = 5.2

**B&B决策：**
1. 初始时，没有找到整数解
2. 解原始LP（不分支）：得到上界，比如 6.0
3. 分支到左右两个子问题
4. 解左分支LP：上界 = 5.6，解到整数解(5,5)，Z = 4.8
5. 解右分支LP：上界 = 4.7
   - 此时已经有整数解Z = 4.8（来自左分支）
   - 右分支上界 4.7 < 4.8
   - **剪掉右分支**，因为不可能在右分支找到比4.8更好的解

**结果：全局最优解是(5,5)，Z = 4.8** ✓

这正是 B&B 应该找到的答案！

### 总结要点

| 方面 | 说明 |
|------|------|
| **上界的含义** | LP最优值是该分支可行域内任何点的objective上界 |
| **为什么有效** | 这是线性规划的定义性质，与目标函数形状无关 |
| **剪枝逻辑** | 如果分支上界 < 当前最优整数解，该分支不可能改进全局最优 |
| **对整数解的影响** | 整数解必然满足该分支的上界，不会跳过最优解 |
| **关键不变量** | 维护当前最优整数解，与分支上界逐个比较 |


## 深入理解：Chvátal-Gomory (CG) 割详解

### 什么是CG割？为什么要乘向量？

CG割的核心思想是：**通过组合多个约束来产生新的约束，这个新约束可能削除LP的分数解而不伤害整数解**。

### 简单例子：从直观到数学

#### 问题设置
```
max Z = x₁ + 1.5x₂
s.t. 3x₁ + 2x₂ ≤ 12    (约束1)
     x₁ + 2x₂ ≤ 8       (约束2)
     x₁, x₂ ≥ 0, 整数
```

#### 第1步：解LP松弛（无整数约束）
- LP最优解：(2.4, 2.4)
- Z_LP = 2.4 + 1.5×2.4 = 5.4
- **问题**：这不是整数解

#### 第2步：选择乘数向量并组合约束

这是关键步骤！假设我们选择 $u = [u_1, u_2] = [0.5, 0.5]$（这个选择有讲究！）

**用u去乘以两个约束：**

原始约束组合：
$$0.5 \times (3x_1 + 2x_2 \leq 12) + 0.5 \times (x_1 + 2x_2 \leq 8)$$

计算左边：
$$0.5 \times 3x_1 + 0.5 \times x_1 + 0.5 \times 2x_2 + 0.5 \times 2x_2$$
$$= 1.5x_1 + 1.5x_1 + x_2 + x_2$$
$$= 2x_1 + 2x_2$$

计算右边：
$$0.5 \times 12 + 0.5 \times 8 = 6 + 4 = 10$$

所以得到：
$$2x_1 + 2x_2 \leq 10$$

#### 第3步：向下取整（这是魔法！）

**为什么可以向下取整？** 因为 $x_1$ 和 $x_2$ 都是整数，所以 $2x_1 + 2x_2$ 也必然是整数！

如果一个整数 ≤ 10，那它也 ≤ ⌊10⌋ = 10。这看起来没变化...

让我换一个例子！假设 $u = [0.4, 0.3]$：

$$0.4 \times (3x_1 + 2x_2 \leq 12) + 0.3 \times (x_1 + 2x_2 \leq 8)$$

左边系数：
- $x_1$: $0.4 \times 3 + 0.3 \times 1 = 1.2 + 0.3 = 1.5$
- $x_2$: $0.4 \times 2 + 0.3 \times 2 = 0.8 + 0.6 = 1.4$

右边：$0.4 \times 12 + 0.3 \times 8 = 4.8 + 2.4 = 7.2$

得到：
$$1.5x_1 + 1.4x_2 \leq 7.2$$

**因为x是整数，左边是整数倍的组合，所以是整数。任何整数都 ≤ 7.2意味着它 ≤ 7**

向下取整得到**CG割**：
$$\lfloor 1.5 \rfloor x_1 + \lfloor 1.4 \rfloor x_2 \leq \lfloor 7.2 \rfloor$$
$$1x_1 + 1x_2 \leq 7$$
$$x_1 + x_2 \leq 7$$

#### 第4步：检查威力

原始LP最优：(2.4, 2.4)
- 代入新约束：2.4 + 2.4 = 4.8 ≤ 7 ✓（满足，但被削减了）

但是这条新约束的威力在于：它定义了一个更紧的可行域！

### 为什么这有效？关键的数学性质

**关键观察 1：为什么能向下取整？**

如果 $x$ 是整数向量，那么 $\lfloor A \rfloor x$ 必然是整数。

证明：设 $A$ 的第 $i$ 行为 $\lfloor a_i \rfloor$，$x$ 为整数
$$\lfloor a_i \rfloor x = \lfloor a_i \rfloor \cdot x \text{ （整数乘以整数）}$$

所以如果 $a_i x \leq b_i$（对所有整数 $x$），那么：
$$\lfloor a_i \rfloor x \leq \lfloor a_i x \rfloor \leq \lfloor b_i \rfloor$$

**关键观察 2：为什么不会削除整数解？**

任何满足原始约束的整数解 $x^*$：
- 满足 $Ax^* \leq b$
- 用 $u$ 乘：$u^T Ax^* \leq u^T b$
- 因为 $x^*$ 是整数，$u^T Ax^*$ 也是整数
- 所以 $\lfloor u^T A \rfloor x^* \leq \lfloor u^T b \rfloor$（等号可能成立或严格不等）
- 新约束满足！✓

**关键观察 3：为什么削除分数解？**

LP最优 $(2.4, 2.4)$ 虽然满足原约束，但：
- 代入新约束：$1 \times 2.4 + 1 \times 2.4 = 4.8$
- 4.8 ≤ 7，仍然满足... 

这个例子不够好。让我重新选择 $u = [0.3, 0.3]$：

$$0.3 \times (3x_1 + 2x_2 \leq 12) + 0.3 \times (x_1 + 2x_2 \leq 8)$$
$$0.9x_1 + 0.6x_2 + 0.3x_1 + 0.6x_2 \leq 3.6 + 2.4$$
$$1.2x_1 + 1.2x_2 \leq 6$$

向下取整：
$$\lfloor 1.2 \rfloor x_1 + \lfloor 1.2 \rfloor x_2 \leq \lfloor 6 \rfloor$$
$$x_1 + x_2 \leq 6$$

现在 $(2.4, 2.4)$ 代入：$2.4 + 2.4 = 4.8 ≤ 6$ 仍满足...

**真正的威力在于多个割的累积！**

### 实际例子：看到真实的削除

```
max Z = 3x₁ + 2x₂
s.t. 10x₁ + 6x₂ ≤ 45
     x₁, x₂ ≥ 0, 整数
```

LP最优：$(4.5, 0)$，$Z = 13.5$

现在用 $u = 0.5$ 乘以约束：
$$0.5 \times (10x_1 + 6x_2 \leq 45)$$
$$5x_1 + 3x_2 \leq 22.5$$

向下取整（因为 $x_1, x_2$ 是整数）：
$$5x_1 + 3x_2 \leq \lfloor 22.5 \rfloor = 22$$

现在检查 LP 最优点 $(4.5, 0)$：
- 原约束：$10 \times 4.5 + 6 \times 0 = 45 ≤ 45$ ✓
- CG割：$5 \times 4.5 + 3 \times 0 = 22.5 ≤ 22$ ✗ **被削除！**

整数解 $(4, 0)$：
- 原约束：$10 \times 4 + 6 \times 0 = 40 ≤ 45$ ✓
- CG割：$5 \times 4 + 3 \times 0 = 20 ≤ 22$ ✓ **仍然满足！**

**这就是CG割的威力：削除分数LP解，保留整数可行解！**

### 算法流程总结

```
【步骤1】给定IP的约束 Ax ≤ b，x为整数

【步骤2】解LP松弛 → 得到分数最优解 x*_LP

【步骤3】选择乘数向量 u ≥ 0
         （这个选择很重要！好的u能产生强割）

【步骤4】计算组合约束：
         u^T A x ≤ u^T b

【步骤5】对系数向下取整：
         ⌊u^T A⌋ x ≤ ⌊u^T b⌋
         
         这就是CG割！

【步骤6】如果新约束被LP最优解违反：
         它削除了x*_LP，但保留了所有整数解
         
【步骤7】将新割加入LP，重新求解 → 更紧的界

【步骤8】重复直到LP最优是整数（或不再找到好的割）
```

### 为什么选择u很关键？

- $u = 0$ → 得不到任何新信息
- $u$ 太小 → 得到弱割，几乎不削除任何东西
- $u$ 好地选择 → 得到强割，大幅削除分数空间

在实际中，选择好的 $u$ 很困难——这也是为什么割平面单独使用时效率不高。现代求解器使用启发式和特殊的割家族（如Gomory割、MIR割等）来自动产生好的割。


In [10]:
print("=" * 100)
print("CG割的实际演示：从线性组合到整数削除")
print("=" * 100)

# 例子：
# max Z = 3x₁ + 2x₂
# s.t. 10x₁ + 6x₂ ≤ 45
#      x₁, x₂ ≥ 0, 整数

print("\n【原始IP问题】")
print("max Z = 3x₁ + 2x₂")
print("s.t. 10x₁ + 6x₂ ≤ 45")
print("     x₁, x₂ ≥ 0, 整数")

# 第1步：解LP松弛
print("\n" + "="*100)
print("【第1步】解LP松弛（去掉整数约束）")
print("="*100)

m_lp = Model("LP Relaxation")
x1_lp = m_lp.addVar(vtype=GRB.CONTINUOUS, name="x1", lb=0)
x2_lp = m_lp.addVar(vtype=GRB.CONTINUOUS, name="x2", lb=0)
m_lp.setObjective(3*x1_lp + 2*x2_lp, GRB.MAXIMIZE)
m_lp.addConstr(10*x1_lp + 6*x2_lp <= 45)
m_lp.optimize()

print(f"LP最优解：x₁ = {x1_lp.X:.2f}, x₂ = {x2_lp.X:.2f}")
print(f"目标值：Z_LP = {m_lp.ObjVal:.2f}")
print(f"✗ 不是整数！")

# 第2步：生成CG割
print("\n" + "="*100)
print("【第2步】生成CG割")
print("="*100)

# 选择乘数向量 u
u = 0.5
print(f"\n选择乘数向量：u = {u}")
print(f"目的：用这个向量乘以所有约束，然后向下取整")

print(f"\n原始约束：10x₁ + 6x₂ ≤ 45")
print(f"用 u = {u} 乘以约束：")
print(f"  {u} × (10x₁ + 6x₂ ≤ 45)")
print(f"  {u*10}x₁ + {u*6}x₂ ≤ {u*45}")
print(f"  {u*10}x₁ + {u*6}x₂ ≤ {u*45}")

# 因为x是整数，左边是整数的倍数
print(f"\n关键观察：因为x₁, x₂是整数")
print(f"  → {u*10}x₁是整数（整数 × 某个数，然后向下取整）")
print(f"  → {u*6}x₂也是整数")
print(f"  → 所以左边的和也是整数")

print(f"\n向下取整右边：")
print(f"  ⌊{u*10}⌋x₁ + ⌊{u*6}⌋x₂ ≤ ⌊{u*45}⌋")
print(f"  {int(np.floor(u*10))}x₁ + {int(np.floor(u*6))}x₂ ≤ {int(np.floor(u*45))}")

cg_coeff1 = int(np.floor(u*10))
cg_coeff2 = int(np.floor(u*6))
cg_rhs = int(np.floor(u*45))

print(f"\n✓ CG割：{cg_coeff1}x₁ + {cg_coeff2}x₂ ≤ {cg_rhs}")

# 第3步：检查LP最优是否被削除
print("\n" + "="*100)
print("【第3步】检查：CG割是否削除了LP最优解？")
print("="*100)

lp_value = cg_coeff1 * x1_lp.X + cg_coeff2 * x2_lp.X
print(f"\nLP最优点 ({x1_lp.X:.2f}, {x2_lp.X:.2f}) 代入CG割：")
print(f"  {cg_coeff1} × {x1_lp.X:.2f} + {cg_coeff2} × {x2_lp.X:.2f} = {lp_value:.2f}")
print(f"  {lp_value:.2f} ≤ {cg_rhs}? ", end="")

if lp_value > cg_rhs:
    print(f"✗ 否！LP最优解被削除了！")
else:
    print(f"✓ 是。LP最优解仍满足。")
    
# 第4步：检查所有整数解是否仍被保留
print("\n" + "="*100)
print("【第4步】检查：整数解是否仍被保留？")
print("="*100)

print(f"\n列出所有整数可行解：")
feasible_solutions = []
for x1_val in range(5):
    for x2_val in range(8):
        if 10*x1_val + 6*x2_val <= 45:
            cg_value = cg_coeff1 * x1_val + cg_coeff2 * x2_val
            satisfies_cg = cg_value <= cg_rhs
            obj_value = 3*x1_val + 2*x2_val
            feasible_solutions.append({
                'x₁': x1_val,
                'x₂': x2_val,
                'Z': obj_value,
                'CG满足': satisfies_cg,
                'CG值': cg_value
            })

df_sol = pd.DataFrame(feasible_solutions).sort_values('Z', ascending=False)
print(f"\n按目标值排序，前10个解：")
print(df_sol.head(10).to_string(index=False))

# 验证最优整数解
best_int = df_sol.iloc[0]
print(f"\n✓ 最优整数解：x₁ = {best_int['x₁']}, x₂ = {best_int['x₂']}, Z = {best_int['Z']}")
print(f"  验证CG割：{best_int['CG值']} ≤ {cg_rhs}? {best_int['CG满足']}")

# 对比：加入CG割后求解
print("\n" + "="*100)
print("【第5步】加入CG割，重新求解LP（模拟削除）")
print("="*100)

m_with_cg = Model("LP with CG cut")
x1_cg = m_with_cg.addVar(vtype=GRB.CONTINUOUS, name="x1", lb=0)
x2_cg = m_with_cg.addVar(vtype=GRB.CONTINUOUS, name="x2", lb=0)
m_with_cg.setObjective(3*x1_cg + 2*x2_cg, GRB.MAXIMIZE)
m_with_cg.addConstr(10*x1_cg + 6*x2_cg <= 45)
m_with_cg.addConstr(cg_coeff1*x1_cg + cg_coeff2*x2_cg <= cg_rhs)  # 添加CG割
m_with_cg.optimize()

print(f"\n新LP最优解：x₁ = {x1_cg.X:.2f}, x₂ = {x2_cg.X:.2f}")
print(f"新目标值：Z_LP = {m_with_cg.ObjVal:.2f}")
print(f"\n对比：")
print(f"  原LP最优：Z = {m_lp.ObjVal:.2f}")
print(f"  加CG割后：Z = {m_with_cg.ObjVal:.2f}")
print(f"  界变紧了 {m_lp.ObjVal - m_with_cg.ObjVal:.2f}")

# 直接求解IP
print("\n" + "="*100)
print("【第6步】对比：直接求解IP")
print("="*100)

m_ip = Model("IP")
x1_ip = m_ip.addVar(vtype=GRB.INTEGER, name="x1", lb=0)
x2_ip = m_ip.addVar(vtype=GRB.INTEGER, name="x2", lb=0)
m_ip.setObjective(3*x1_ip + 2*x2_ip, GRB.MAXIMIZE)
m_ip.addConstr(10*x1_ip + 6*x2_ip <= 45)
m_ip.optimize()

print(f"IP最优解：x₁ = {x1_ip.X:.0f}, x₂ = {x2_ip.X:.0f}")
print(f"IP最优值：Z* = {m_ip.ObjVal:.0f}")

print("\n" + "="*100)
print("【总结】")
print("="*100)
print(f"""
CG割的工作流程：

1. 乘以向量u：将所有约束用u乘起来
   → {u} × (10x₁ + 6x₂ ≤ 45) = {u*10}x₁ + {u*6}x₂ ≤ {u*45}

2. 向下取整：因为x是整数，所以左边是整数
   → ⌊{u*10}⌋x₁ + ⌊{u*6}⌋x₂ ≤ ⌊{u*45}⌋
   → {cg_coeff1}x₁ + {cg_coeff2}x₂ ≤ {cg_rhs}

3. 效果：
   - 原LP最优 ({x1_lp.X:.2f}, {x2_lp.X:.2f}) 被削除
   - 但所有整数解仍被保留
   - 新LP界更紧：从 {m_lp.ObjVal:.2f} → {m_with_cg.ObjVal:.2f}

4. 意义：
   - CG割逐步缩小LP可行域
   - 最终使LP最优解都是整数（或接近）
   - 这样就能有效求解IP！
""")


CG割的实际演示：从线性组合到整数削除

【原始IP问题】
max Z = 3x₁ + 2x₂
s.t. 10x₁ + 6x₂ ≤ 45
     x₁, x₂ ≥ 0, 整数

【第1步】解LP松弛（去掉整数约束）
Gurobi Optimizer version 13.0.1 build v13.0.1rc0 (mac64[arm] - Darwin 24.6.0 24G517)

CPU model: Apple M4 Pro
Thread count: 14 physical cores, 14 logical processors, using up to 14 threads

Optimize a model with 1 rows, 2 columns and 2 nonzeros (Max)
Model fingerprint: 0xa8c19507
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [6e+00, 1e+01]
  Objective range  [2e+00, 3e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [4e+01, 4e+01]

Presolve removed 1 rows and 2 columns
Presolve time: 0.00s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.5000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.00 seconds (0.00 work units)
Optimal objective  1.500000000e+01
LP最优解：x₁ = 0.00, x₂ = 7.50
目标值：Z_LP = 15.00
✗ 不是整数！

【第2步】生成CG割

选择乘数向量：u = 0.5
目的：用这个向

## 关键问题：整数解从何而来？

您刚才指出的问题非常敏锐：**如果我们只求LP松弛，哪里来的"当前最佳整数解"？**

### B&B 完整流程中的整数解来源

整数解可以从以下几个途径获得：

#### 方法1：在分支树的叶子节点获得（最常见）
当某个分支的LP最优解恰好是整数时，我们就得到了一个IP可行解：

```
原始问题 (分支)
    ↓
LP松弛：最优解 (4.7, 5.3) ← 不是整数
    ↓
分支1: x_f ≤ 5          分支2: x_f ≥ 6
    ↓                      ↓
LP最优: (4, 5) ✓       LP最优: (2, 6) ✓
是整数！发现可行解       是整数！发现可行解
Z = 26                   Z = 24
```

#### 方法2：启发式算法（工业求解器内部）
现代求解器（如Gurobi）在B&B过程中会：
- 使用启发式算法寻找可行整数解（不需要遍历整个分支树）
- 这给了我们一个初始的"当前最佳整数解"
- 例如：贪心、四舍五入、局部搜索等

#### 方法3：直接求解某个分支为整数程序
在某些分支上，直接解整数版本而不只是LP松弛，更快地找到整数解

### B&B 的正确完整步骤

让我展示一个现实的B&B过程：

```
【第0步】原始问题（LP松弛）
  max Z = 3x_e + 2x_f
  s.t. x_e + 2x_f ≤ 12
  LP最优：(4, 2) = 16
  
  ↓ 这是一个整数！我们找到了一个可行解！
  
  当前最佳整数解：Z* = 16，点为(4, 2)
  当前上界：16
  
【第1步】对x_e分支（因为在其他分支上可能有更优解）
  
  分支A: x_e ≤ 3           分支B: x_e ≥ 4
      ↓                         ↓
      LP: (3, 4.5)          LP: (4, 2)
      Z_LP = 15             Z_LP = 16
      ↓                         ↓
      15 < 16?是                16 = 16?是，但已是整数
      剪掉这个分支！          这就是目前的最优
      
结论：B&B搜索完成，全局最优解为 Z* = 16，(4, 2)
```

### 为什么刚才例子中有"当前最佳整数解"

在我之前的演示中，我假设了某个时刻已经找到了一个整数解。这可能通过以下方式：

1. **启发式找到**：在解LP之前或之后，求解器先快速找到一个整数解
2. **从某个分支发现**：在B&B树的某个地方，某个分支的LP最优恰好是整数
3. **显式求解**：如果直接调用IP求解器（而不是手动B&B），求解器内部会完整执行B&B，自然会产生整数解

### 实际情况：当没有整数解时

如果LP松弛的最优解（在某个分支）是非整数的，而我们尚未找到任何整数解：

```
【假设】
分支1的LP最优：(4.7, 5.3)，Z_LP = 24.5
分支2的LP最优：(2, 6)，Z_LP = 18（是整数）

步骤：
1. 先解分支2（假设）
   → 得到整数解 Z* = 18
   
2. 再解分支1
   → LP最优 24.5 > 18
   → 需要继续分支
   
   分支1a: x_f ≤ 5 → LP最优 (4, 5)，Z = 17（整数！）
   → 发现更好的整数解 Z* = 17
   
3. 回到分支1的其他子分支
   → 都被 Z* = 17 剪掉
```

### 关键理解

| 阶段 | 操作 |
|------|------|
| **初始化** | 使用启发式或简单方法找一个初始整数解（初始 Z*），或 Z* = -∞ |
| **分支** | 选择一个节点，解其LP松弛 |
| **检查** | LP最优是否为整数？|
| **如果整数** | 发现新可行解，更新 Z*（如果更优） |
| **如果非整数** | 继续分支 |
| **剪枝** | 若 LP上界 ≤ 当前 Z*，剪掉该分支 |
| **终止** | 所有分支都被剪掉或探索完，返回最后的 Z* |

**最关键的点**：B&B 不仅是解LP，而是通过不断分支来：
1. **发现整数解**（通过LP偶然是整数，或启发式）
2. **改进上界**（通过分支缩小可行域）
3. **证明最优性**（通过证明不存在更好的可能）


In [9]:
print("=" * 90)
print("完整的B&B过程演示：从哪里找到整数解？")
print("=" * 90)

# 构建一个例子，其中LP松弛的最优不是整数
print("\n【原始IP】")
print("max Z = 3*x_e + 2*x_f")
print("s.t. x_e + 2*x_f ≤ 12")
print("     x_e, x_f ≥ 0, 整数")

# 第0步：解原始LP（没有整数约束）
print("\n" + "="*90)
print("【第0步：解原始LP松弛（无分支）】")
print("="*90)

m_root = Model("Root LP")
x_e_root = m_root.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0)
x_f_root = m_root.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)
m_root.setObjective(3*x_e_root + 2*x_f_root, GRB.MAXIMIZE)
m_root.addConstr(x_e_root + 2*x_f_root <= 12)
m_root.optimize()

print(f"LP最优解：x_e = {x_e_root.X:.1f}, x_f = {x_f_root.X:.1f}")
print(f"目标值：Z_LP = {m_root.ObjVal:.1f}")
print(f"是整数吗？{x_e_root.X == int(x_e_root.X) and x_f_root.X == int(x_f_root.X)}")

best_integer_solution = None
best_integer_value = -float('inf')

# 如果LP最优已经是整数，立即找到解
if x_e_root.X == int(x_e_root.X) and x_f_root.X == int(x_f_root.X):
    best_integer_solution = (x_e_root.X, x_f_root.X)
    best_integer_value = m_root.ObjVal
    print(f"✓ 找到整数解：({best_integer_solution[0]:.0f}, {best_integer_solution[1]:.0f})，Z = {best_integer_value:.1f}")
else:
    print("✗ LP最优不是整数，需要分支")

# 第1步：在x_e分支
print("\n" + "="*90)
print("【第1步：在 x_e 上分支】")
print("="*90)

# 分支1：x_e ≤ 4（向下取整）
print("\n【分支1：x_e ≤ 4】")
m1 = Model("Branch 1: x_e <= 4")
x_e1 = m1.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=0, ub=4)
x_f1 = m1.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)
m1.setObjective(3*x_e1 + 2*x_f1, GRB.MAXIMIZE)
m1.addConstr(x_e1 + 2*x_f1 <= 12)
m1.optimize()

print(f"  LP最优：x_e = {x_e1.X:.1f}, x_f = {x_f1.X:.1f}")
print(f"  目标值：Z_LP = {m1.ObjVal:.1f}")
print(f"  是整数吗？{x_e1.X == int(x_e1.X) and x_f1.X == int(x_f1.X)}")

if x_e1.X == int(x_e1.X) and x_f1.X == int(x_f1.X):
    print(f"  ✓ 找到整数解：({x_e1.X:.0f}, {x_f1.X:.0f})，Z = {m1.ObjVal:.1f}")
    if m1.ObjVal > best_integer_value:
        best_integer_solution = (x_e1.X, x_f1.X)
        best_integer_value = m1.ObjVal
        print(f"  更新最佳整数解！Z* = {best_integer_value:.1f}")

# 分支2：x_e ≥ 5（向上取整）
print("\n【分支2：x_e ≥ 5】")
m2 = Model("Branch 2: x_e >= 5")
x_e2 = m2.addVar(vtype=GRB.CONTINUOUS, name="x_e", lb=5)
x_f2 = m2.addVar(vtype=GRB.CONTINUOUS, name="x_f", lb=0)
m2.setObjective(3*x_e2 + 2*x_f2, GRB.MAXIMIZE)
m2.addConstr(x_e2 + 2*x_f2 <= 12)
m2.optimize()

print(f"  LP最优：x_e = {x_e2.X:.1f}, x_f = {x_f2.X:.1f}")
print(f"  目标值：Z_LP = {m2.ObjVal:.1f}")
print(f"  是整数吗？{x_e2.X == int(x_e2.X) and x_f2.X == int(x_f2.X)}")

if x_e2.X == int(x_e2.X) and x_f2.X == int(x_f2.X):
    print(f"  ✓ 找到整数解：({x_e2.X:.0f}, {x_f2.X:.0f})，Z = {m2.ObjVal:.1f}")
    if m2.ObjVal > best_integer_value:
        best_integer_solution = (x_e2.X, x_f2.X)
        best_integer_value = m2.ObjVal
        print(f"  更新最佳整数解！Z* = {best_integer_value:.1f}")

# 剪枝检查
print("\n" + "="*90)
print("【剪枝决策】")
print("="*90)

if best_integer_value > -float('inf'):
    print(f"当前最佳整数解：Z* = {best_integer_value:.1f}")
    print(f"\n分支1：LP上界 = {m1.ObjVal:.1f}")
    if m1.ObjVal > best_integer_value:
        print(f"  {m1.ObjVal:.1f} > {best_integer_value:.1f} → 保留分支，可能找到更优解")
    else:
        print(f"  {m1.ObjVal:.1f} ≤ {best_integer_value:.1f} → 剪掉分支")
    
    print(f"\n分支2：LP上界 = {m2.ObjVal:.1f}")
    if m2.ObjVal > best_integer_value:
        print(f"  {m2.ObjVal:.1f} > {best_integer_value:.1f} → 保留分支，可能找到更优解")
    else:
        print(f"  {m2.ObjVal:.1f} ≤ {best_integer_value:.1f} → 剪掉分支")
else:
    print("还未找到任何整数解，继续分支...")

# 对比：直接求解IP
print("\n" + "="*90)
print("【对比：直接用Gurobi求解IP（内部完整B&B）】")
print("="*90)

m_ip = Model("Direct IP")
x_e_ip = m_ip.addVar(vtype=GRB.INTEGER, name="x_e", lb=0)
x_f_ip = m_ip.addVar(vtype=GRB.INTEGER, name="x_f", lb=0)
m_ip.setObjective(3*x_e_ip + 2*x_f_ip, GRB.MAXIMIZE)
m_ip.addConstr(x_e_ip + 2*x_f_ip <= 12)
m_ip.optimize()

print(f"IP最优解：x_e = {x_e_ip.X:.0f}, x_f = {x_f_ip.X:.0f}")
print(f"目标值：Z* = {m_ip.ObjVal:.0f}")

print("\n" + "="*90)
print("【总结】")
print("="*90)
print(f"""
1. 当我们解原始LP时，它给出一个上界：Z ≤ {m_root.ObjVal:.1f}

2. 如果LP最优是整数，我们直接找到IP的解
   → 这时"当前最佳整数解"就来自这个LP

3. 如果LP最优不是整数，我们分支：
   → 每个分支也求LP松弛
   → 如果某分支的LP最优是整数，就是一个可行解
   → 用这个可行解作为"当前最佳整数解"

4. 用"当前最佳整数解"与其他分支的上界对比：
   → 上界更高的分支继续探索
   → 上界更低的分支剪掉

在这个例子中：
  - 根节点LP不是整数（{m_root.ObjVal:.1f}）
  - 分支1的LP是整数：({x_e1.X:.0f}, {x_f1.X:.0f})，Z = {m1.ObjVal:.1f}
  - 分支2的LP是整数：({x_e2.X:.0f}, {x_f2.X:.0f})，Z = {m2.ObjVal:.1f}
  - 最好的是分支1，所以 Z* = {max(m1.ObjVal, m2.ObjVal):.1f}

这与直接求解IP的结果一致：Z* = {m_ip.ObjVal:.0f}
""")


完整的B&B过程演示：从哪里找到整数解？

【原始IP】
max Z = 3*x_e + 2*x_f
s.t. x_e + 2*x_f ≤ 12
     x_e, x_f ≥ 0, 整数

【第0步：解原始LP松弛（无分支）】
Gurobi Optimizer version 13.0.1 build v13.0.1rc0 (mac64[arm] - Darwin 24.6.0 24G517)

CPU model: Apple M4 Pro
Thread count: 14 physical cores, 14 logical processors, using up to 14 threads

Optimize a model with 1 rows, 2 columns and 2 nonzeros (Max)
Model fingerprint: 0xe4d2dbf0
Model has 2 linear objective coefficients
Coefficient statistics:
  Matrix range     [1e+00, 2e+00]
  Objective range  [2e+00, 3e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+01, 1e+01]

Presolve removed 1 rows and 2 columns
Presolve time: 0.01s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    3.6000000e+01   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.01 seconds (0.00 work units)
Optimal objective  3.600000000e+01
LP最优解：x_e = 12.0, x_f = 0.0
目标值：Z_LP = 36.0
是整数吗？True
✓ 找到整数解：(12, 0)，Z = 36.0

【